lib/Conversion/ExportVerilog/ExportVerilog.cpp

```diff
@@ -4046,6 +3978,134 @@ isExpressionEmittedInlineIntoProceduralDeclaration(Operation *op,
   return true;
 }
 
+template <class AssignTy>
+static AssignTy getSingleAssignAndCheckUsers(Operation *op) {
+  AssignTy singleAssign;
+  if (llvm::all_of(op->getUsers(), [&](Operation *user) {
+        if (hasSVAttributes(user))
+          return false;
+
+        if (auto assign = dyn_cast<AssignTy>(user)) {
+          if (singleAssign)
+            return false;
+          singleAssign = assign;
+          return true;
+        }
+
+        return isa<ReadInOutOp>(user);
+      }))
+    return singleAssign;
+  return {};
+}
+
+// Return true if `op1` dominates users of `op2`.
+static bool checkDominanceOfUsers(Operation *op1, Operation *op2) {
+  return llvm::all_of(op2->getUsers(), [&](Operation *user) {
+    if (op1->getBlock() != user->getBlock())
```

Comment:
I guess this is conservative in the case of nested regions, and that could be made more precise using MLIR's DominanceInfo? Seems like a good place to start for now since it's simple, and isBeforeInBlock is actually amortized to constant time on average. We can always compute DominanceInfo for whole modules later if more and more things care about this.

Reply:
Right, I agree that eventually we want to use DominanceInfo here.